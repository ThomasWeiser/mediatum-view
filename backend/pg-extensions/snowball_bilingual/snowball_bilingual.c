
/*-------------------------------------------------------------------------
 *
 * snowball_bilingual.c
 *	  Snowball dictionary as extension
 *
 * Copyright (c) 2007-2019, PostgreSQL Global Development Group
 *
 * 2019 Modified for bilinguality by MediaTUM project, Thomas Weiser
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "commands/defrem.h"
#include "tsearch/ts_locale.h"
#include "tsearch/ts_utils.h"

/* Now we can include the original Snowball header.h */
#include "snowball/libstemmer/header.h"
#include "libstemmer/stem_UTF_8_english.h"
#include "libstemmer/stem_UTF_8_german.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(dsnowball_bilingual_init);
PG_FUNCTION_INFO_V1(dsnowball_bilingual_lexize);

/* List of supported modules */
typedef struct stemmer_module
{
	const char *name;
	pg_enc enc;
	struct SN_env *(*create)(void);
	void (*close)(struct SN_env *);
	int (*stem)(struct SN_env *);
} stemmer_module;

static const stemmer_module stemmer_modules[] =
		{
				// Stemmers list from Snowball distribution
				{"english", PG_UTF8, english_UTF_8_create_env, english_UTF_8_close_env, english_UTF_8_stem},
				{"german", PG_UTF8, german_UTF_8_create_env, german_UTF_8_close_env, german_UTF_8_stem},

				{NULL, 0, NULL, NULL, NULL} /* list end marker */
};

typedef struct DictSnowballLanguage
{
	struct SN_env *z;
	bool needrecode; /* needs recoding before/after call stem */
	int (*stem)(struct SN_env *z);
	StopList stoplist;
} DictSnowballLanguage;

typedef struct DictSnowballBilingual
{
	DictSnowballLanguage languages[2];

	/*
	 * snowball saves alloced memory between calls, so we should run it in our
	 * private memory context. Note, init function is executed in long lived
	 * context, so we just remember CurrentMemoryContext
	 */
	MemoryContext dictCtx;
} DictSnowballBilingual;

static void
locate_stem_module(DictSnowballLanguage *dl, char *lang)
{
	const stemmer_module *m;

	readstoplist(lang, &dl->stoplist, lowerstr);

	/*
	 * First, try to find exact match of stemmer module. Stemmer with
	 * PG_SQL_ASCII encoding is treated as working with any server encoding
	 */
	for (m = stemmer_modules; m->name; m++)
	{
		if ((m->enc == PG_SQL_ASCII || m->enc == GetDatabaseEncoding()) &&
				pg_strcasecmp(m->name, lang) == 0)
		{
			dl->stem = m->stem;
			dl->z = m->create();
			dl->needrecode = false;
			return;
		}
	}

	/*
	 * Second, try to find stemmer for needed language for UTF8 encoding.
	 */
	for (m = stemmer_modules; m->name; m++)
	{
		if (m->enc == PG_UTF8 && pg_strcasecmp(m->name, lang) == 0)
		{
			dl->stem = m->stem;
			dl->z = m->create();
			dl->needrecode = true;
			return;
		}
	}

	ereport(ERROR,
					(errcode(ERRCODE_UNDEFINED_OBJECT),
					 errmsg("no Snowball stemmer available for language \"%s\" and encoding \"%s\"",
									lang, GetDatabaseEncodingName())));
}

Datum
		dsnowball_bilingual_init(PG_FUNCTION_ARGS)
{
	List *dictoptions = (List *)PG_GETARG_POINTER(0);
	DictSnowballBilingual *db;
	ListCell *l;

	// elog(WARNING, "sizeof(DictSnowballLanguage) = %lu", sizeof(DictSnowballLanguage));
	// elog(WARNING, "sizeof(DictSnowballBilingual) = %lu", sizeof(DictSnowballBilingual));

	db = (DictSnowballBilingual *)palloc0(sizeof(DictSnowballBilingual));

	foreach (l, dictoptions)
	{
		DefElem *defel = (DefElem *)lfirst(l);

		if (pg_strcasecmp("StopWords", defel->defname) == 0)
		{
			ereport(ERROR,
							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
							 errmsg("unexpected StopWords parameter. This bilingual Snowball template is fixed to English and German.")));
		}
		else if (pg_strcasecmp("Language", defel->defname) == 0)
		{
			ereport(ERROR,
							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
							 errmsg("unexpected Language parameter. This bilingual Snowball template is fixed to English and German.")));
		}
		else
		{
			ereport(ERROR,
							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
							 errmsg("unrecognized Snowball parameter: \"%s\"",
											defel->defname)));
		}
	}

	locate_stem_module(&(db->languages[0]), "english");
	locate_stem_module(&(db->languages[1]), "german");

	db->dictCtx = CurrentMemoryContext;

	PG_RETURN_POINTER(db);
}

static void
lexize_language(TSLexeme *res, char *txt, DictSnowballLanguage *dl, MemoryContext dictCtx)
{
	MemoryContext saveCtx;

	/*
	 * recode to utf8 if stemmer is utf8 and doesn't match server encoding
	 */
	if (dl->needrecode)
	{
		char *recoded;

		recoded = pg_server_to_any(txt, strlen(txt), PG_UTF8);
		if (recoded != txt)
		{
			pfree(txt);
			txt = recoded;
		}
	}

	/* see comment about db->dictCtx */
	saveCtx = MemoryContextSwitchTo(dictCtx);
	SN_set_current(dl->z, strlen(txt), (symbol *)txt);
	dl->stem(dl->z);
	MemoryContextSwitchTo(saveCtx);

	if (dl->z->p && dl->z->l)
	{
		txt = repalloc(txt, dl->z->l + 1);
		memcpy(txt, dl->z->p, dl->z->l);
		txt[dl->z->l] = '\0';
	}

	/* back recode if needed */
	if (dl->needrecode)
	{
		char *recoded;

		recoded = pg_any_to_server(txt, strlen(txt), PG_UTF8);
		if (recoded != txt)
		{
			pfree(txt);
			txt = recoded;
		}
	}

	res->lexeme = txt;
}

Datum
		dsnowball_bilingual_lexize(PG_FUNCTION_ARGS)
{
	DictSnowballBilingual *db = (DictSnowballBilingual *)PG_GETARG_POINTER(0);
	char *in = (char *)PG_GETARG_POINTER(1);
	int32 len = PG_GETARG_INT32(2);
	TSLexeme *res = palloc0(sizeof(TSLexeme) * 3);

	char *txt = lowerstr_with_len(in, len);

	if (*txt == '\0')
	{
		pfree(txt);
	}
	else
	{
		int lexemeCount = 0;

		if (searchstoplist(&(db->languages[0].stoplist), txt))
		{
			pfree(txt);
		}
		else
		{
			lexize_language(res + (lexemeCount++), txt, &(db->languages[0]), db->dictCtx);
		}

		// TODO: Don't copy the string again. Needs some memory management refactoring.
		txt = lowerstr_with_len(in, len);

		if (searchstoplist(&(db->languages[1].stoplist), txt))
		{
			pfree(txt);
		}
		else
		{
			lexize_language(res + (lexemeCount++), txt, &(db->languages[1]), db->dictCtx);
		}

		if (lexemeCount == 2)
		{
			// Mark the second lexeme a variant of the first lexeme.
			// This is relevant for queries to combine the lexemes by `|` instead of `&`
			// (i.e. disjuction instead of conjunction).
			res[1].nvariant = 1;

			// If the lexemes are equal then just keep one version
			if (strcmp(res[0].lexeme, res[1].lexeme) == 0)
			{
				memset(res + 1, 0, sizeof(TSLexeme));
			}
		}
	}

	PG_RETURN_POINTER(res);
}
