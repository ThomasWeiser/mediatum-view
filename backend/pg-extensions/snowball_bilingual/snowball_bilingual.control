# snowball_bilingual extension
comment = 'snowball_bilingual - add-on dictionary template using the combined results of two snowball stemmers'
default_version = '1.0'
module_pathname = '$libdir/snowball_bilingual'
relocatable = true
