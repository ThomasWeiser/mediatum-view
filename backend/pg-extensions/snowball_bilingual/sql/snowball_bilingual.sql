CREATE EXTENSION snowball_bilingual;

SELECT * FROM ts_debug('english_german','Welt Welten world worlds');

SELECT to_tsvector('english_german', 'Welt Welten world worlds');
SELECT to_tsquery('english_german', 'Welten');

CREATE TABLE table1(name varchar);
INSERT INTO table1 VALUES
	('I''m sorry'),
	('Ich bitte um Entschuldigung'),
	('PostgreSQL the world''s most advanced open source databases'),
	('PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt'),
	('i me my myself the'),
	('als also am an der die das');
SELECT d.* FROM table1 AS t, LATERAL ts_debug('english_german', t.name) AS d;

CREATE INDEX name_idx ON table1 USING GIN (to_tsvector('english_german', "name"));

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sorry');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'entschuldigen');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sources');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'Welten');

DROP INDEX name_idx;

CREATE INDEX name_idx ON table1 USING GIST (to_tsvector('english_german', "name"));

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sorry');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'entschuldigen');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sources');
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'Welten');

DROP INDEX name_idx;
