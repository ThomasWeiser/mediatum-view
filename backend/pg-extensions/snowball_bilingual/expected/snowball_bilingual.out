CREATE EXTENSION snowball_bilingual;
SELECT * FROM ts_debug('english_german','Welt Welten world worlds');
   alias   |   description   | token  |     dictionaries      |     dictionary      |    lexemes    
-----------+-----------------+--------+-----------------------+---------------------+---------------
 asciiword | Word, all ASCII | Welt   | {english_german_stem} | english_german_stem | {welt}
 blank     | Space symbols   |        | {}                    |                     | 
 asciiword | Word, all ASCII | Welten | {english_german_stem} | english_german_stem | {welten,welt}
 blank     | Space symbols   |        | {}                    |                     | 
 asciiword | Word, all ASCII | world  | {english_german_stem} | english_german_stem | {world}
 blank     | Space symbols   |        | {}                    |                     | 
 asciiword | Word, all ASCII | worlds | {english_german_stem} | english_german_stem | {world}
(7 rows)

SELECT to_tsvector('english_german', 'Welt Welten world worlds');
            to_tsvector            
-----------------------------------
 'welt':1,2 'welten':2 'world':3,4
(1 row)

SELECT to_tsquery('english_german', 'Welten');
    to_tsquery     
-------------------
 'welten' | 'welt'
(1 row)

CREATE TABLE table1(name varchar);
INSERT INTO table1 VALUES
	('I''m sorry'),
	('Ich bitte um Entschuldigung'),
	('PostgreSQL the world''s most advanced open source databases'),
	('PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt'),
	('i me my myself the'),
	('als also am an der die das');
SELECT d.* FROM table1 AS t, LATERAL ts_debug('english_german', t.name) AS d;
      alias      |           description           |            token            |     dictionaries      |     dictionary      |                         lexemes                         
-----------------+---------------------------------+-----------------------------+-----------------------+---------------------+---------------------------------------------------------
 asciiword       | Word, all ASCII                 | I                           | {english_german_stem} | english_german_stem | {i}
 blank           | Space symbols                   | '                           | {}                    |                     | 
 asciiword       | Word, all ASCII                 | m                           | {english_german_stem} | english_german_stem | {m}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | sorry                       | {english_german_stem} | english_german_stem | {sorri,sorry}
 asciiword       | Word, all ASCII                 | Ich                         | {english_german_stem} | english_german_stem | {ich}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | bitte                       | {english_german_stem} | english_german_stem | {bitt}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | um                          | {english_german_stem} | english_german_stem | {um}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | Entschuldigung              | {english_german_stem} | english_german_stem | {entschuldigung,entschuld}
 asciiword       | Word, all ASCII                 | PostgreSQL                  | {english_german_stem} | english_german_stem | {postgresql}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | the                         | {english_german_stem} | english_german_stem | {the}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | world                       | {english_german_stem} | english_german_stem | {world}
 blank           | Space symbols                   | '                           | {}                    |                     | 
 asciiword       | Word, all ASCII                 | s                           | {english_german_stem} | english_german_stem | {s}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | most                        | {english_german_stem} | english_german_stem | {most}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | advanced                    | {english_german_stem} | english_german_stem | {advanc,advanced}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | open                        | {english_german_stem} | english_german_stem | {open}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | source                      | {english_german_stem} | english_german_stem | {sourc}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | databases                   | {english_german_stem} | english_german_stem | {databas}
 asciiword       | Word, all ASCII                 | PostgreSQL                  | {english_german_stem} | english_german_stem | {postgresql}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | das                         | {english_german_stem} | english_german_stem | {das}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | fortschrittlichste          | {english_german_stem} | english_german_stem | {fortschrittlichst,fortschritt}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciihword      | Hyphenated word, all ASCII      | Open-Source-Datenbanksystem | {english_german_stem} | english_german_stem | {open-source-datenbanksystem,open-source-datenbanksyst}
 hword_asciipart | Hyphenated word part, all ASCII | Open                        | {english_german_stem} | english_german_stem | {open}
 blank           | Space symbols                   | -                           | {}                    |                     | 
 hword_asciipart | Hyphenated word part, all ASCII | Source                      | {english_german_stem} | english_german_stem | {sourc}
 blank           | Space symbols                   | -                           | {}                    |                     | 
 hword_asciipart | Hyphenated word part, all ASCII | Datenbanksystem             | {english_german_stem} | english_german_stem | {datenbanksystem,datenbanksyst}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | der                         | {english_german_stem} | english_german_stem | {der}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | Welt                        | {english_german_stem} | english_german_stem | {welt}
 asciiword       | Word, all ASCII                 | i                           | {english_german_stem} | english_german_stem | {i}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | me                          | {english_german_stem} | english_german_stem | {me}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | my                          | {english_german_stem} | english_german_stem | {my}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | myself                      | {english_german_stem} | english_german_stem | {myself}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | the                         | {english_german_stem} | english_german_stem | {the}
 asciiword       | Word, all ASCII                 | als                         | {english_german_stem} | english_german_stem | {al}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | also                        | {english_german_stem} | english_german_stem | {also}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | am                          | {english_german_stem} | english_german_stem | {}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | an                          | {english_german_stem} | english_german_stem | {}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | der                         | {english_german_stem} | english_german_stem | {der}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | die                         | {english_german_stem} | english_german_stem | {die}
 blank           | Space symbols                   |                             | {}                    |                     | 
 asciiword       | Word, all ASCII                 | das                         | {english_german_stem} | english_german_stem | {das}
(67 rows)

CREATE INDEX name_idx ON table1 USING GIN (to_tsvector('english_german', "name"));
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sorry');
   name    
-----------
 I'm sorry
(1 row)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'entschuldigen');
            name             
-----------------------------
 Ich bitte um Entschuldigung
(1 row)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sources');
                                  name                                  
------------------------------------------------------------------------
 PostgreSQL the world's most advanced open source databases
 PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt
(2 rows)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'Welten');
                                  name                                  
------------------------------------------------------------------------
 PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt
(1 row)

DROP INDEX name_idx;
CREATE INDEX name_idx ON table1 USING GIST (to_tsvector('english_german', "name"));
SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sorry');
   name    
-----------
 I'm sorry
(1 row)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'entschuldigen');
            name             
-----------------------------
 Ich bitte um Entschuldigung
(1 row)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'sources');
                                  name                                  
------------------------------------------------------------------------
 PostgreSQL the world's most advanced open source databases
 PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt
(2 rows)

SELECT * FROM table1 WHERE to_tsvector('english_german', name)
	@@ to_tsquery('english_german', 'Welten');
                                  name                                  
------------------------------------------------------------------------
 PostgreSQL das fortschrittlichste Open-Source-Datenbanksystem der Welt
(1 row)

DROP INDEX name_idx;
